packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx

```diff
@@ -1,449 +1,209 @@
-  getRowProps?: (id: GridRowId, model: GridRowModel) => any;
-}
+export type VirtualScroller = ReturnType<typeof useGridVirtualScroller>;
 
-interface ContainerDimensions {
-  width: number | null;
-  height: number | null;
-}
-
-// The `maxSize` is 3 so that reselect caches the `renderedColumns` values for the pinned left,
-// unpinned, and pinned right sections.
-const MEMOIZE_OPTIONS = { maxSize: 3 };
-
-export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
+export const useGridVirtualScroller = () => {
   const apiRef = useGridPrivateApiContext();
   const rootProps = useGridRootProps();
   const visibleColumns = useGridSelector(apiRef, gridVisibleColumnDefinitionsSelector);
   const enabled = useGridSelector(apiRef, gridVirtualizationEnabledSelector);
   const enabledForColumns = useGridSelector(apiRef, gridVirtualizationColumnEnabledSelector);
-
-  const {
-    ref,
-    onRenderZonePositioning,
-    renderZoneMinColumnIndex = 0,
-    renderZoneMaxColumnIndex = visibleColumns.length,
-    getRowProps,
-  } = props;
+  const dimensions = useGridSelector(apiRef, gridDimensionsSelector);
+  const outerSize = dimensions.viewportOuterSize;
+  const innerSize = dimensions.viewportInnerSize;
+  const pinnedRows = useGridSelector(apiRef, gridPinnedRowsSelector);
+  const pinnedColumns = useGridSelector(apiRef, gridVisiblePinnedColumnDefinitionsSelector);
+  const hasBottomPinnedRows = pinnedRows.bottom.length > 0;
+  const [panels, setPanels] = React.useState(EMPTY_DETAIL_PANELS);
 
   const theme = useTheme();
   const columnPositions = useGridSelector(apiRef, gridColumnPositionsSelector);
-  const columnsTotalWidth = useGridSelector(apiRef, gridColumnsTotalWidthSelector);
   const cellFocus = useGridSelector(apiRef, gridFocusCellSelector);
   const cellTabIndex = useGridSelector(apiRef, gridTabIndexCellSelector);
   const rowsMeta = useGridSelector(apiRef, gridRowsMetaSelector);
   const selectedRowsLookup = useGridSelector(apiRef, selectedIdsLookupSelector);
   const currentPage = useGridVisibleRows(apiRef, rootProps);
+  const gridRootRef = apiRef.current.rootElementRef;
+  const mainRef = apiRef.current.mainElementRef;
+  const scrollerRef = apiRef.current.virtualScrollerRef;
   const renderZoneRef = React.useRef<HTMLDivElement>(null);
-  const rootRef = React.useRef<HTMLDivElement>(null);
-  const handleRef = useForkRef(ref, rootRef);
-  const [renderContext, setRenderContextState] = React.useState<GridRenderContext | null>(null);
-  const prevRenderContext = React.useRef<GridRenderContext | null>(renderContext);
-  const scrollPosition = React.useRef({ top: 0, left: 0 });
-  const [containerDimensions, setContainerDimensions] = React.useState<ContainerDimensions>({
-    width: null,
-    height: null,
-  });
+  const scrollbarVerticalRef = React.useRef<HTMLDivElement>(null);
+  const scrollbarHorizontalRef = React.useRef<HTMLDivElement>(null);
+  const contentHeight = dimensions.contentSize.height;
```

Comment:
There is a bit of extra space at the bottom in some cases, like here: https://deploy-preview-10059--material-ui-x.netlify.app/x/react-data-grid/column-pinning/


https://github.com/mui/mui-x/assets/13808724/c1a85096-dcde-4557-8956-bd6d3a102acd

Reply:
Logic issue with the render context, the final row wasn't rendered. It's fixed.